sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+    pub fn execute_command(
+        &self,
+        cmd: &str,
+        mut callback: impl FnMut(&str)) -> Result<()>
+    {
+        use std::io::Read;
+        use std::io::Write;
+        let mut port = serialport::new(self.console_tty.to_str().ok_or(Error::UnicodePathError)?, 115_200)
+            .timeout(std::time::Duration::from_millis(100))
+            .open().expect("Failed to open port");
+
+        // Ideally, we would invoke Linux flock() on the serial
+        // device, to detect minicom or another instance of
+        // opentitantool having the same serial port open.  Incoming
+        // serial data could go silenly missing, in such cases.
+        let mut buf = [0u8; 128];
+        loop {
+            match port.read(&mut buf[0..128]) {
+                Ok(rc) => {
+                    log::info!("Discarded {} characters: {:?}\n", rc,
+                               &std::str::from_utf8(&buf[0..rc]));
+                }
+                Err(error) => {
+                    match error.kind() {
+                        std::io::ErrorKind::TimedOut => {
+                            break;
+                        }
+                        _ => {
+                            return Err(error.into())
+                        }
+                    }
+                }
+            }
+        }
+        // Send Ctrl-C, followed by the command, then newline.  This will discard any previous
+        // partial input, before executing our command.
+        port.write(format!("\u{3}{}\n", cmd).as_bytes())?;
+        let mut seen_echo = false;
+        let mut len: usize = 0;
+        loop {
+            match port.read(&mut buf[len..128]) {
+                Ok(rc) => {
+                    len += rc;
+                    'outer: loop {
+                        for i in 0..len {
+                            if buf[i] == 10 {
+                                // Newline
+                                let mut end = i;
+                                if end > 0 && buf[end - 1] == 13 {
+                                    end -= 1;
+                                }
+                                let line = std::str::from_utf8(&buf[0..end])?;
+                                if seen_echo {
+                                    callback(line);
+                                } else {
+                                    if line.len() >= cmd.len()
+                                        && line[line.len() - cmd.len()..] == *cmd {
+                                        seen_echo = true;
+                                    }
+                                }
+                                buf.rotate_left(i + 1);
+                                len -= i + 1;
+                                continue 'outer;
+                            }
+                        }
+                        break
+                    }
+                }
+                Err(error) if error.kind() == std::io::ErrorKind::TimedOut => {
```

Comment:
ErrorKind is used a lot around here... `use` it?

Reply:
Done.